utils/embedded-test-support/avr-qemu-atmega2560/support.c

```diff
@@ -0,0 +1,62 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors.
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <stddef.h>
+#include <stdint.h>
+
+__attribute__((naked))
+__attribute__((section(".start")))
+void start() {
+  // stack is 0x1e00 ..< 0x2200 (1 kB), see linkerscript.ld
+  // set SP to 0x21ff, last byte of SRAM
+  asm volatile("ldi r16, 0xff");
+  asm volatile("out 0x3d, r16");
+  asm volatile("ldi r16, 0x21");
+  asm volatile("out 0x3e, r16");
+  asm volatile("call copy_data_from_flash_to_sram");
+  asm volatile("call usart_init");
+  asm volatile("call main");
+  asm volatile("call halt");
+}
+
+void usart_init() {
+  *((volatile char *)0xc1) = 1 << 3; // enable TX on UART0
+}
+
+__attribute__((used))
+static inline void *memcpy_flash_to_sram(void *restrict dst, const void __attribute__((__address_space__(1))) *restrict src, size_t n) {
+  for (int i = 0; i < n; i++) {
+    ((char *)dst)[i] = ((char __attribute__((__address_space__(1))) *)src)[i];
+  }
+  return dst;
+}
+
+void copy_data_from_flash_to_sram() {
+  // Copy data segment from 0x200-program-space to 0x200-data-space, 7 kB in size (see linkerscript.ld)
+  memcpy_flash_to_sram((void *)0x200, (void __attribute__((__address_space__(1))) *)0x200, 7 * 1024);
+}
+
+int putchar(int c) {
+  // AVR's UART0 DR register
+  *((volatile char *)0xc6) = c;
```

Comment:
nit: This should work fine in QEMU although it probably wouldn't work on a device, as you need to wait for the byte to transmit through the UART, something like: https://github.com/microswift-packages/HAL/blob/develop/module/UART.swift#L472 (we got caught out by this a few times as swift often optimised this check away!)

Reply:
FWIW: that repo is private and access through a volatile pointer should never be optimized away. Could you share the source?

Reply:
(it's written in swift as we use pure swift for our hardware libraries these days but you get the idea)

Reply:
hmm you should use swift-mmio and the svd2swift code generator once more parts of the swift compiler/swiftpm build system are working